### PR TITLE
basic video creation interface

### PR DESCRIPTION
**Pull request by Velro**
Tested on Windows and Ubuntu. Drop an ffmpeg executable into the root
folder of the project or add it to your PATH. I am not committing the
executable to the repo to keep size down (its about 40mb * number of
platforms).

The debugVideo.mp4 is to show that it can append new framest o an
existing video. Ifyou don't specify an "appendee" video it'll create a
video from the frames you give it. This is all outlined in the code.

The following command should generate a video

node ffmpegWrapper.js "resources/images" 1 "resources/debugVideo.mp4"

### CODE REVIEW

**Comment by Velro** [comment]
Ehsan, I'm not sure how to get the pictures from S3 to this through Lambda. Here is the logic I'm thinking would work from doing a similar kind of project with text message images.

1. New images go to a folder called "pending" or something
2. Lambda is notified either with each new image, or when a certain threshold of images in the folder is reached.
3.  download the current video.
4. downloads all images in the "pending" folder
5. Append the new frames to the old video.
6. Upload new video.
7. Move images on S3 from "pending" folder to a "complete" folder

**Comment by rimutaka** [comment]
Thanx Jason. We are trying to make it work today.

File: ffmpegWrapper.js [hunk 1]
```diff
@@ -76,13 +76,79 @@ function addAudio()
     fs.renameSync("finalVideoNoAudio.mp4", "finalVideo.mp4");
 }
 
+function timestampFilenameCompare(var a, var b)
```
Comment (eafzali) [inline ffmpegWrapper.js]: 'var a'? in function signature?

File: ffmpegWrapper.js [hunk 2]
```diff
@@ -76,13 +76,79 @@ function addAudio()
...
+{
+    var indexOfTA = a.indexOf("T");
+    var dateA = parseInt(a.slice(0, indexOfTA));
+    var timeOfDayA = parseInt(a.slice(indexOfTA + 1, a.IndexOf(".")));
```
Comment (eafzali) [inline ffmpegWrapper.js]: IndexOf => indexOf

File: ffmpegWrapper.js [hunk 3]
```diff
@@ -76,13 +76,79 @@ function addAudio()
...
+                    console.log("File in image directory isn't a jpg");
+                    return;
+                }
+            }
+
+            fileNames.sort(timestampFilenameCompare);
+
+            for (var fileIndex = 0; fileIndex < fileNames.length; fileIndex++)
+            {
+                var newName = util.format('%03d.jpg', fileIndex);
```
Comment (eafzali) [inline ffmpegWrapper.js]: util is not imported!

**Comment by eafzali** [comment]
Hi @Velro, I guess you didn't run this code at all. Would you please clean it up, run it once and commit again?

File: ffmpegWrapper.js (hunk 1 above)
Comment (eafzali) [inline ffmpegWrapper.js]: What is the need of this compare function? simple string compare function (that is the default one used by sort function when you don't specify any compare function) will yield the same result on ISO datetime strings.

**Comment by Velro** [comment]
Hi Ehsan, thank you for your feedback. I accidently uploaded the incorrect file, that was an in-progress version. I will update again to conform to linting shortly.

File: ffmpegWrapper.js [hunk 4]
```diff
@@ -17,186 +17,171 @@
     existing video must be an mp4
 */
 
-'use strict';
```
Comment (eafzali) [inline ffmpegWrapper.js]: Why did you remove it?

File: ffmpegWrapper.js [hunk 5]
```diff
@@ -17,186 +17,171 @@
...
+  NOTE(james): ffmpeg needs a list of files to be concatenated, so the output
+  of this file is something like
 
-    var concatListContent = "";
-    concatListContent += "file '"+ existingVideo + "'\r\n"; //NOTE(james): becomes - file '/path/to/file1'
-    concatListContent += "file 'newFramesVideoOutput.mp4'\r\n";
+  file 'existingVideo.mp4'
+  file 'newFramesVideoOuput.mp4'
 
-    fs.writeFileSync("concatList.txt", concatListContent);
```
Comment (eafzali) [inline ffmpegWrapper.js]: Do not use sync operations. and write all your intermediate files on `/tmp`.

File: ffmpegWrapper.js [hunk 6]
```diff
@@ -17,186 +17,171 @@
...
+// NOTE(james): if main hasn't been set, like I think it would be in AWS,
+// then we are using the offline CLI
+if (require.main === module) {
+  offlineCLI();
+}
+/*
+// NOTE(james): AWS Path
+exports.handler = (event, context, callback) =>
 {
-    if (process.platform == "win32")
```
Comment (eafzali) [inline ffmpegWrapper.js]: We won't need windows platform support for sure so you can get rid of all checkes. just assume that ffmpeg binary is there in the project folder.

File: ffmpegWrapper.js [hunk 7]
```diff
@@ -17,186 +17,171 @@
...
-    var concatListContent = "";
-    concatListContent += "file '"+ existingVideo + "'\r\n"; //NOTE(james): becomes - file '/path/to/file1'
-    concatListContent += "file 'newFramesVideoOutput.mp4'\r\n";
+  file 'existingVideo.mp4'
+  file 'newFramesVideoOuput.mp4'
 
-    fs.writeFileSync("concatList.txt", concatListContent);
+  The concatenation file is deleted at the end of this program
+*/
+  const concatList = `$file \'${existingVideo}\'\r\nfile \'newFramesVideoOutput.mp4\'\r\n`;
```
Comment (eafzali) [inline ffmpegWrapper.js]: That `$file` should be `file`?

**Comment by eafzali** [comment]
Thanks @Velro seems much better. Please consider we are going to import your code as a module and call a function of it like this:

```js
const ffmpeg = require('./ffmpegWrapper.js');

ffmpeg.convert(imagesFolder, fps, prevVideo) //prevVideo would be optional
.then(pathToVideo => { console.log('Final video is here', pathToVideo) })
```

So please export a convert function that returns a promise and can be used like above example. 

Some important things you should follow so that your code would run on AWS Lambda correctly:

1. Please don't use sync IO commands at all.
2. Always use `/tmp` to store your intermediate files and always use a UUID for the name, because that folder will be shared between different instances and they may overwrite each other.

**Comment by Velro** [comment]
Hey @eafzali I started switching to async IO calls and returning as a Promise. I have included a new file that just tests the import, we can delete it before merging. I'm having some trouble wrapping my head around promises, and am having trouble with them. I have a couple spots labelled HELP, could you please give me some feedback? Particularly, the call site in what would be your module never goes into the .then(...) block, but the ffmpegWrapper function completed successfully. I've been struggling with it for an hour or so, seems like something that you would understand quickly.

Also you mentioned that temporary files need to have a UUID name. I have implemented that, but should the final video name be one as well? If more than one instance is creating videos I would think we would hit the same conflict problem.

**Comment by rimutaka** [comment]
There should be only one instance of this function at a time. The file name
should be in the config.

**Comment by eafzali** [comment]
@rimutaka what about multiple cams? 
@Velro Yes use an UUID for the final video as well and that's why I required you to return it's name at the end. I will push it to s3 with proper name afterward.

**Comment by rimutaka** [comment]
Different cams live in different folders.
Is there a problem? Did I miss something?

**Comment by eafzali** [comment]
On S3 yes, but we don't have the folder structure on Lambda, we have to use `/tmp` to store all intermediate files. I'll create a temp folder in `/tmp` download new images and prev video into it. run the ffmpeg and then upload the new video into it's place on s3. makes sense?

**Comment by rimutaka** [comment]
It does now. You know what you are doing. I'll leave it for you to resolve.

**Comment by Velro** [comment]
I am working on it now, should have it in an hour or so. This should be the final pass.

**Comment by Velro** [comment]
Okay I have it working how you requested. I also switched it so the param paths are relative to tmp, since it seems most of the work will be happening in there. Should be trivial for you to change if needed. I still have the one weird sync IO point that I'm unsure how to replace, when I rename the image files.

**Comment by rimutaka** [comment]
James, do you have any sample videos to share?
There are plenty of images on S3. Should be enough for a few seconds of 
video.

**Comment by Velro** [comment]
1 frame per second video https://drive.google.com/open?id=0B1dDx93UEQfDZHByTEhqcmpKdzQ
4 frames per second video https://drive.google.com/open?id=0B1dDx93UEQfDalVScXFhaTFBZVU

I'd upload directly but github doesn't support it. I've attached a zip file with them too.
[exampleVideos.zip](https://github.com/hubsy-io/timelapse/files/475935/exampleVideos.zip)

The black and white intro numbers is just a random video for testing the append functionality. This is using 13 of your images.

**Comment by eafzali** [comment]
OK @Velro I have already taken care of that rename thing I'm renaming all image files on fly when downloading them from s3 so you can remove that part and assume that the image files are in the required format.

Would you please create another pull request agains current HEAD of master branch and just fill https://github.com/hubsy-io/timelapse/blob/master/functions/video-encoder/src/ffmpeg.js
we don't need all these images and videos to be committed to the repo.

**Comment by Velro** [comment]
It looks like you took care of everything. Do you need me to do anything
else?

**Comment by rimutaka** [comment]
James, Ehsan ran into vi generation problems. It was generating videos 
48kb long and crashing. He suspects the source images are too big.

**Comment by eafzali** [comment]
yes @Velro it seems that ffmpeg crashes in the middle of the process on lambda but I confirm that it's working on my mac, I've posted the problem here:

http://ffmpeg.org/pipermail/ffmpeg-user/2016-September/033661.html

**Comment by eafzali** [comment]
@Velro seems that problem was about the order of ffmpeg args. I moved '-y' to the end before output file name and it worked!

**Comment by Velro** [comment]
Weird. Glad you got it working though!
